### ISSUE TITLE
Frontlight plugin on Android should be disabled

### ISSUE THREAD

**Issue by gerroon** [koreader/koreader#3118]
Hi

I really think that the frontlight plugin or whatever plugin that is trying to control the frontlight on Android should be disabled. It is nothing but a pain in the neck for me.

It always start "black" on  all of my devices. Even the latest release from today comes up black, until I pull down the system menu when Iget my screen. THen I haeve to fiddle around to force refresh, do monkey magic etc to set it up. The worst case for me is when Koreader crashes in the middle of the night when I am reading, then I restart Koreader and boom no light, I do not know what I am looking at, so I just turn off Koreader and goto sleep instead of keep reading ;( This is the worst issue I have using Koreader on Inkbook.

Koreader should not try to overwrite the system light, for instance it disables "auto brightness" on my phone then it maxes out the light on the device. The result is blinding brightness in the middle of the night, and it stays that way even I quit Koreader.

Is there a way to disable all these mind reading frontlight actions? If not can you please disable this functionality on Android until this is figured out for real.

thanks

**Comment by Frenzie** [comment]
@poire-z Is this part of #3103?

**Comment by poire-z** [comment]
No, I don't plan touching something I can't test :|

But may be it's just a matter of testing changing this to `no` ? :
https://github.com/koreader/koreader/blob/abaa699080e6af8c2ebb08cb21b452bb2cf6b1b3/frontend/device/android/device.lua#L14

**Comment by Frenzie** [comment]
I can test it to some degree on my old Xperia Ray phone (KOReader doesn't work on my Zenfone 2). KOReader seems intent on changing the brightness to some halfway value when it starts. I don't know about E Ink devices, but on that phone it's somewhere in between odd and potentially annoying because KOReader takes control and the system's frontlight control no longer works.

So yeah, if E Ink Android devices are basically the same as far as frontlight control goes then there would be little point to KOReader controlling the frontlight in the first place.

**Comment by gerroon** [comment]
When I start Koreader on my Inkbook, I have no light also the screen always comes up black (on all my Android deviuces). On my eink, it should not disable the light, like I said it sux in the middle of the night when I need to restart Koreader, given that it Koreader tends to crash sometimes.

When I start Koreader on my phone, it disables auto broghtness of the device and does some highbrightness whcih I do not understand. Why does it mess with the system light? It should just remember the last Koreader session's light level at best.

**Comment by Frenzie** [comment]
If I understand correctly then the frontlight control added when fixing #1415 was fine and some more recent change made it odd?

**Comment by gerroon** [comment]
Yes I think so, at least it got worse last couple months. The situation was fine until last couple months but I might be wrong because multiple issues are happening on various devices of mine.

**Comment by Frenzie** [comment]
@gerroon So you're saying this is mysteriously better now?

**Comment by bahusoid** [comment]
@Frenzie Why do you think mysteriously? I saw some specific code regarding it https://github.com/koreader/android-luajit-launcher/pull/62/files:
```
if brightness > 255 then
     brightness = 255
 elseif brightness < 0 then
     brightness = 0
end
```
I personally think this code needs to be updated to quit this routine in case brightness  level is incorrect.

**Comment by Frenzie** [comment]
@bahusoid Nah, that shouldn't stop KOReader from changing any values, just from requesting nonsensical ones for whatever reason.

I noticed @Hzj-jie committed https://github.com/koreader/koreader/commit/12af617b468809e2b2b2eaed91511c9fe95af98d which might be the key. But no reference was made to this issue or #3004.

Actually for me the current build doesn't seem to be capable of controlling the frontlight *at all*, which I don't mind (and it nicely avoids any potential issues) but it would seem to be a bug.

**Comment by Frenzie** [comment]
Yeah, I don't understand the change pointed out by @bahusoid  because the minimal frontlight level KOReader will ever call on Android from the GUI is already 0 and the maximum is already 255 (i.e., 255*25/25). It's an extra safety check that doesn't hurt but also shouldn't be needed.

The inability to actually set the frontlight level is caused by the addition of this check in #3254:
https://github.com/koreader/koreader/blob/0fc5a378bc6282325d39b67de552b56309acdb8a/frontend/device/android/powerd.lua#L14

However, it looks like without that line KOReader might do that "starting black" thing.

I'm attaching a build without that check so you can see if it does anything weird with the frontlight.

[koreader-android-arm-linux-androideabi-v2015.11-1255-g0fc5a378-frontlight-fix.apk.zip](https://github.com/koreader/koreader/files/1338336/koreader-android-arm-linux-androideabi-v2015.11-1255-g0fc5a378-frontlight-fix.apk.zip)

If it doesn't do anything weird then we can look into simply making it ignore stuff on startup. If it does all kinds of weird stuff it might be better to leave it purposefully broken for the time being.

**Comment by bahusoid** [comment]
@Frenzie It seems "Black" screen issue is not related to frontlight. As I already [mentioned ](https://github.com/koreader/koreader/issues/3214#issuecomment-330029699) for me black screen was caused by attempt to change brightness to 2601. Or brightness and frontlight are the same things, are they?

So I just think we should have an option to disable  brightness manipulations. And in my opinion current fix should be changed to something like (not sure if it's valid lua, just a pseudocode):
```
if (brightness > 255 or brightness < 0) then
	--do not set invalid brightness value, as we can't be sure what
	--this value really represent, so let's just quit from setScreenBrightness 
     return
```

**Comment by Frenzie** [comment]
@bahusoid That's not what I'm seeing.

```
09-27 20:40:48.584 14394 14407 I KOReader: set screen brightness 153
```

It's just that KOReader is black until you touch it and force a redraw, and it's triggered by that single line I mentioned above. The real question is why it's trying to set that value on startup at all imho. It's just getting it and then setting it.

Anyway, your pseudocode is exactly what is done now thanks to the change you pointed out? I'm not going to remove that. It just doesn't serve any purpose as it already shouldn't ever try to use such values in principle. I have no idea how you could've possibly gotten 2601 out of it. :-P

**Comment by bahusoid** [comment]
@Frenzie  Did you check attached log in message I referred? It's an app start - I didn't touch anything on UI. I just installed app and started it.
[NoVersionedCrashLog.txt](https://github.com/koreader/koreader/files/1308947/NoVersionedCrashLog.txt)
```

09-17 11:33:42.440 I/luajit-launcher( 3292): Screen size 1072x1448
09-17 11:33:42.440 I/luajit-launcher( 3292): Screen brightness 255
...
09-17 11:33:43.610 I/KOReader( 3292): set screen brightness 2601
```
So yeah I also not sure where did KOReader get it. But it's not me : )

**Comment by Frenzie** [comment]
I found some suspicious stuff:

https://github.com/koreader/koreader/blob/0fc5a378bc6282325d39b67de552b56309acdb8a/frontend/device/generic/powerd.lua#L24-L26

https://github.com/koreader/koreader/blob/0fc5a378bc6282325d39b67de552b56309acdb8a/frontend/device/generic/powerd.lua#L148-L157

I don't really understand why KOReader should maintain a different frontlight state from the rest of the system — regardless whether we're talking Kobo, Android or anything else.

**Comment by bahusoid** [comment]
@Frenzie I just checked your 1255-g0fc5a378-frontlight-fix - and it's starting again with black screen for me (frontlight is still ON, it's just black screen)

**Comment by Frenzie** [comment]
I made a stupid mistake. The black screen seems to be caused by the patch.lua thing that I accidentally still had active. (Like I said, it'll go away if you tap the screen.) What I don't understand is why it does this, because I don't even have a patch.lua.

Oddly enough it also seems to affect the frontlight somehow.

None of this makes any sense. :-P

**Comment by bahusoid** [comment]
Well - it didn't go away for me quickly on tap. I made Hystory view to be open on start (not sure if it's important). And it took me quite some time to see it (not sure what exactly I did to make it go away - but it took me minutes of taps and swipes).

**Comment by Frenzie** [comment]
The pcall issue seems to go away if I change the incantation to:
```lua
pcall(dofile, "/sdcard/koreader/patch.lua")
```

(instead of using an anonymous function `pcall(function() dofile("/sdcard/koreader/patch.lua") end)`)

Now I'm not sure why that should make a difference, but if we like the patch.lua thing it's good to know.

**Comment by bahusoid** [comment]
What's all the fuss about patch.lua? :) Where can I read about? As I have no idea why would I need it  :)

**Comment by Frenzie** [comment]
You shouldn't need it, but you can use it to apply some overrides for testing (or regular use) if desired.

1. Android loads luajit-launcher
2. luajit-launcher prepares necessary stuff
3. Main KOReader UI loads

But just before 3 there's the possibility for a patch.lua override, such as:

```lua
local screen = require("android").screen
screen.height = screen.height + 45
```
But you can probably also do wilder things, like

```lua
local android = require("android")
android.setScreenBritghtnessOrig = android.setScreenBrightness
-- set brightness to highest value no matter what, muhahahaha
android.setScreenBrightness = function() android.setScreenBritghtnessOrig(255) end
```

**Comment by Frenzie** [comment]
@poire-z This (open, brightness-related) issue seems like a more appropriate location. If you get rid of these lines I pointed out above then you don't see the problem.

https://github.com/koreader/koreader/blob/0fc5a378bc6282325d39b67de552b56309acdb8a/frontend/device/generic/powerd.lua#L24-L26

What you notice then is that the problem (black screen) occurs when you first set the frontlight.

I have no idea what's going on though.

A simple workaround is to stick `self.device.screen:refreshFull()` after this line.

https://github.com/koreader/koreader/blob/34ad91c3f0baa2ca21bf6add35c1ce6bb0297bbe/frontend/device/android/powerd.lua#L13-L15

I think a `Screen:refreshFull()` might work somewhere sufficiently late (where I gave it a quick try in powerd and uimanager I got some `Unable to lock window buffer`) as a simple workaround but I'm sure the underlying problem must be something more Android-y.

/me eyes https://github.com/koreader/android-luajit-launcher/blob/19779beabfbd6f8195efcaa998a739f39742c71e/src/org/koreader/launcher/MainActivity.java#L65-L86 suspiciously. ;-)

**Comment by poire-z** [comment]
Could be that this `public void setScreenBrightness(final int brightness) {runOnUiThread(new Runnable() { public void run() {...` runs in a thread, and there is concurrent access to the screen buffer.
What if you add  `require("ffi/util").sleep(2)` after `android.setScreenBrightness `in
```
function AndroidPowerD:setIntensityHW(intensity)
    android.setScreenBrightness(math.floor(255 * intensity / self.fl_max))
end
```

**Comment by Frenzie** [comment]
It doesn't, but there's no reason the brightness should do anything to the window in the first place imo.

Commenting out these three lines solves the problem completely:
https://github.com/koreader/android-luajit-launcher/blob/19779beabfbd6f8195efcaa998a739f39742c71e/src/org/koreader/launcher/MainActivity.java#L78-L80

I can stick it in a commit so we can see how it goes; reverts are easy if there are unintended side effects.

**Comment by poire-z** [comment]
I'll check on tomorrow's nightly. Can you still change the level with the frontlight widget ? :)
(Just read a few page about that, not clear why these 2 things - the `Settings.System.putInt` stuff, which is supposed to set (or just save?) the system level brightness - and the `Window.setAttributes(LayoutParams)` stuff, which sets the app window brightness - are needed, but they go in pair on most - but not all - pages).

**Comment by Frenzie** [comment]
> Can you still change the level with the frontlight widget ? :)

Of course, that's what I meant by solving the problem. ;-)

**Comment by gerroon** [comment]
I just checked the built from 5th and the black screen and the light mess up seems to be back (Lineage OS 7.1). I am not sure if this is just an in progress issue but I thought that I would report it.

**Comment by poire-z** [comment]
The black screen does not happen to me anymore with today's build (v2015.11-1296-gb483c54).

**Comment by gerroon** [comment]
Ok I will check it out

**Comment by poire-z** [comment]
Btw, if you have "auto brightness" enabled, and you don't want koreader to touch light on start (at startup, it queries the current brightness, and set it to that same value - and for setting it, it disables the auto brightness), you can add in `patch.lua` :
```lua
local android = require("android")
android.setScreenBrightnessOrig = android.setScreenBrightness
android.setScreenBrightness = function()
    android.setScreenBrightness = android.setScreenBrightnessOrig
end
```

That way, the first call is a no-op (it just puts back the original function), and subsequent calls will work (and disable auto brightness, so it can set light).

**Comment by Frenzie** [comment]
> at startup, it queries the current brightness, and set it to that same value - and for setting it, it disables the auto brightness

I don't really understand why it's done that way though. I prefer the behavior effected by your patch.lua.

**Comment by poire-z** [comment]
(The generic/powerd stuff looks indeed quite twisted, but there may be a reason for it, to make something generic around which we can deal with devices that we can query light state, those we can't (kobo), and may be for the autofrontlight plugin too - anyway, I spent too much time fixing kobo stuff, that I don't want to revisit it :)

For android, we could just bring the stuff from this patch in device/android/powerd.lua, with something like
```lua
--- a/frontend/device/android/powerd.lua
+++ b/frontend/device/android/powerd.lua
@@ -4,6 +4,7 @@ local _, android = pcall(require, "android")
 local AndroidPowerD = BasePowerD:new{
     fl_min = 0, fl_max = 25,
     fl_intensity = 10,
+    init_done = false,
 }

 function AndroidPowerD:frontlightIntensityHW()
@@ -11,6 +12,12 @@ function AndroidPowerD:frontlightIntensityHW()
 end

 function AndroidPowerD:setIntensityHW(intensity)
+    if not self.init_done then
+        -- Ignore the first set (done by generic init), so we don't
+        -- disable auto-brightness
+        self.init_done = true
+        return
+    end
     android.setScreenBrightness(math.floor(255 * intensity / self.fl_max))
 end
```
(Feel free to test/PR it if that makes sense, as I can't test it)

**Comment by Frenzie** [comment]
Sure, the only reason I didn't do that already is because:

1. I don't understand why the behavior exists.
2. It only displaced the black screen issue anyway.

> those we can't (kobo)

I don't really see what difference it makes frankly. Just leave the darned thing alone unless the user asks for it. I imagine this may have caused the (imo undesirable) behavior of enabling my frontlight all the time which was present for a while. It didn't bug me enough to look at the exact cause, but unless one explicitly enables something like autofrontlight or requests a brightness through the frontlight widget I think the brightness/frontlight shouldn't be touched at all. On any device.

**Comment by gerroon** [comment]
Ok I have an issue with the latest builds on Inkbook. The book always open with no light even though the system has the light set to dimmest already and also the book normally has light when I close it. That is not the issue with my Nougat and Lollipop although I hate that it messes with the auto light, I will try the auto liught solution that @poire-z recommended in an early reply.

**Comment by Frenzie** [comment]
Right, I fixed the frontlight control in https://github.com/koreader/koreader/commit/bf61e6f0932761b413c7451932e2e9af7441481b (five days ago) because it was broken but it comes attached with that weird behavior. Do those builds also do that or do they behave differently? There were two changes not in the same build.

In v2015.11-1254-g6275453 it's completely broken but it seems to me that can more accurately be said about this weird taking control of brightness behavior.

**Comment by poire-z** [comment]
What was happening/not happeing before https://github.com/koreader/koreader/commit/bf61e6f0932761b413c7451932e2e9af7441481b again?
(Looks to me the lines that were removed would may be just be the thing that would nos mess up with light on startup.)

**Comment by Frenzie** [comment]
Absolutely nothing. The self.fl_intensity was set prior to that function so that test was always false. But those lines were only there for like a week.

**Comment by poire-z** [comment]
> I think the brightness/frontlight shouldn't be touched at all. On any device.

You're a no frontlight person, so that wouldn't bother you :) but on kobo:
- boot (frontlight off)
- KSM (fixed, configurable: startWithLightOn (yes for me) and frontLightLevel (4, low, just enough to see and not light the room when plugged to usb for me), but fixed)
- Nickel (sets frontlight to remembered previous level in Kobo ereader.conf - usually 30 or 50 for reading for me)
- KSM again (no ability to read previous level, so again, fixed values => 4)
- Koreader (sets frontlight to remembered previous level in Kobo ereader.conf or else, I expect 30 or 50)

So "_leave the darned thing alone_" would let me at 4 on kobo. You could say "fix KSM, so it goes fetch previous light level in Kobo ereader.conf" so we don't have to, but we're nice guys :)
Or we could just add a new attribute to BasePowerd { leave_the_darn_thing_alone = true } , which would be false on Kobo :)
On all other devices, there's a thing before that has set the light to a remembered state (Android system, Kindle app, Window manager for the emulator). On kobo, there is only KSM. (Note that I have only KSM0.7, may be something changed with 0.8).

> The self.fl_intensity was set prior to that function so that test was always false

Why or how ?

```lua
local AndroidPowerD = BasePowerD:new{    fl_intensity = 10, } -- dummy value, overriden early
function BasePowerD:new(o)
   [... init ...] -- o.fl_intensity = 10
    if o.device and o.device.hasFrontlight() then
        o.fl_intensity = o:frontlightIntensityHW() -- => o.fl_intensity = current android brightness
        o:_decideFrontlightState() -- no-op
        if o:isFrontlightOn() then -- no-op
            o:turnOnFrontlightHW() -- => (*) self:setIntensityHW(self.fl_intensity), which should result in the test being true !
        else
            o:turnOffFrontlightHW()
        end
```
((*): unless at this point, `self` is not yet `o` and self.fl_intensity is not the o.fl_intensity we just set ?!)

**Comment by Frenzie** [comment]
Did I say it the wrong way around on my phone? Whatever the case with true/false, my point is that function was always a no-op because self.fl_intensity == passed intensity and hence frontlight control was non-functional.

But no, if `self.fl_intensity == intensity` then `self.fl_intensity ~= intensity` is `false` ergo I said it right.

**Comment by poire-z** [comment]
just corrected my true/false mismatch :)

**Comment by Frenzie** [comment]
Why would it "need" that?

**Comment by Frenzie** [comment]
Umm sorry but this makes no sense. :-P

```lua
local AndroidPowerD = BasePowerD:new{    fl_intensity = 10, } -- dummy value, overriden early
function BasePowerD:new(o)
   [... init ...] -- o.fl_intensity = 10
    if o.device and o.device.hasFrontlight() then
        o.fl_intensity = o:frontlightIntensityHW() -- => o.fl_intensity = current android brightness
-- so now self.fl_intensity is the real deal, not that inital value of 10
        o:_decideFrontlightState() -- no-op
-- this usually returns true
        if o:isFrontlightOn() then -- no-op
-- ergo so does this
            o:turnOnFrontlightHW() -- => (*) self:setIntensityHW(self.fl_intensity), which should result in the test being false in this call, and so in the light not being changed on startup !
-- so here we get exactly what I said (which also happens from "userland")
-- self.fl_intensity is passed and turned into intensity inside the function
-- but this is not really a problem, although it is absurdly unclear
-- it only really turns into a problem later on when using the frontlight control widget
        else
            o:turnOffFrontlightHW()
        end
```

**Comment by poire-z** [comment]
To work on android today, with the current state of the whole powerd stuff that you don't want to tackle with before your next holidays :)

**Comment by Frenzie** [comment]
If I understand correctly this entire turnOnFrontlight stuff should be Kobo-exclusive?

**Comment by poire-z** [comment]
I don't know :) Guess someone (ask git blame who to ask :) wanted to make a generic flow for initialisation, dunno if it's kobo exclusive, could be with Kobo ereader.conf being the "hardware state".

**Comment by Frenzie** [comment]
Try typing this: `git blame frontend/device/generic/powerd.lua` ;-)

(Also [on GitHub](https://github.com/koreader/koreader/blame/d23d01643edec990e2494ff0c029bd6d3d9e55c7/frontend/device/generic/powerd.lua).)

It's all from #2941.

**Comment by poire-z** [comment]
I guess self.fl_intensity is the current intensity known by powerd, and given to widgets that wants to know.
The original author decided in setIntensity(intensity) to update self.fl_intensity = intensity before calling the hardware methods. May be it could have been updated after that (so HW methods could know previous intensity from self.fl_intensity), but that's not the case. So the use of self.hw_intensity in KoboPowerd (which somehow follows self.fl_intensity)...

**Comment by Frenzie** [comment]
Anyway, the only purpose adding something like that would have is to prevent needless tinkering on startup, which should be prevented globally in the first place rather than being worked around for each and every implementation.

If I understand correctly the solution is this:
```diff
diff --git a/frontend/device/generic/powerd.lua b/frontend/device/generic/powerd.lua
index 2ea4021b..803cf6e2 100644
--- a/frontend/device/generic/powerd.lua
+++ b/frontend/device/generic/powerd.lua
@@ -21,10 +21,12 @@ function BasePowerD:new(o)
     if o.device and o.device.hasFrontlight() then
         o.fl_intensity = o:frontlightIntensityHW()
         o:_decideFrontlightState()
-        if o:isFrontlightOn() then
-            o:turnOnFrontlightHW()
-        else
-            o:turnOffFrontlightHW()
+        if self.device:isKobo() then
+            if o:isFrontlightOn() then
+                o:turnOnFrontlightHW()
+            else
+                o:turnOffFrontlightHW()
+            end
         end
     end
     return o

```

**Comment by Frenzie** [comment]
Actually no, I rather mean that should be moved over to `KoboPowerD:init()`.

**Comment by Frenzie** [comment]
@gerroon Please let me know how tomorrow's nightly works out for you. :-)

**Comment by gerroon** [comment]
Thnx for the lively discussion. I will try the newer builds.

@Frenzie the frontlight not being on was an issue only on Inkbook Prime, my other devices did not have the same issue just as a side note.


Also I would like to iterate what I would like as a die hard koreader fan boy when it comes to light stuff. Not sure if this is  how it is supposed to work but here it goes.

**Step 1** Do not mess with the system's front light when opening or closing the book

**Step 2** Do not change the light settings (turn off, do some average light condition in koreader etc) in koreader if the user "never set the light" in the current book. 

**Step 3** Remember the light setting per book if the user changed the light condition while reading the book. So next time the user opens the book, koreader opens it with  this setting that was set previously within the reading session.

**Step 4** If the user toggles (turns it off) the light in koreader, let the system light setting to be dominant next time the user opens the book. This is tricky because the user may never want the light to be on when the user opens the book next time but if the user opens the book at night, if koreader opens the book with no light, the user will probably just put koreader aside and goto sleep unless the user has some easy access to room or table light.

**Comment by Frenzie** [comment]
1. Should be done now and I agree completely.
2. Sure, same thing.
3. Intriguing thought but nothing like that is currently implemented. 
4. First I've heard of someone without a little nightstand and a light on it or something like that. :-P

**Comment by gerroon** [comment]
@Frenzie 

I am glad that you are with  me on all points except the bed side light ;) I do not have a  plae to put light, I have to walk to another room to turn on koreader's light in these cases, because I do not want to wake my wife up ;)

**Comment by bahusoid** [comment]
@Frenzie No more blackouts for me on app start. But I still can reproduce blackout on first wake up.

So I start Nook, open some book and move device to sleep. When I wake up Nook - I see black screen. 

Interesting part it's reproducible only once per app start. On next wake up - it's OK. But if I restart the app - I can reproduce it again.
[AppStartSleepAndWakeUpWithBlackScreen.txt](https://github.com/koreader/koreader/files/1367850/AppStartSleepAndWakeUpWithBlackScreen.txt)

**Comment by Frenzie** [comment]
@bahusoid That sounds very similar to what I solved on my device by https://github.com/koreader/android-luajit-launcher/pull/64

This is the kind of thing that would really profit from someone who actually knows (and maybe even likes) Android. :-P

**Comment by Frenzie** [comment]
Fixed by #3319.
